### PR TITLE
Add Docs.hasdoc

### PR DESCRIPTION
**Pull request by jariji**
Add @stevengj's `Docs.hasdoc` from https://github.com/JuliaLang/julia/issues/51174.

`check_documented` still should be added but that can happen separately.

### CODE REVIEW

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -655,4 +655,22 @@ function parsedoc end
...
 
+
+"""
+    Docs.hasdoc(mod::Module, sym::Symbol)::Bool
+
+Return `true` if `sym` in `mod` has a docstring and `false` otherwise.
+"""
+hasdoc(mod::Module, sym::Symbol) = hasdoc(Docs.Binding(mod, sym))
+function hasdoc(binding::Docs.Binding, sig::Type = Union{})
+    defined(binding) && !isnothing(getdoc(resolve(binding), sig)) && return true
```
Comment (stevengj) [inline base/docs/Docs.jl:666]:
```suggestion
    # this function is based on the Base.Docs.doc method implemented
    # in REPL/src/docview.jl.  TODO: refactor and unify these methods.
    defined(binding) && !isnothing(getdoc(resolve(binding), sig)) && return true
```

File: test/docs.jl [hunk 2]
```diff
@@ -69,6 +69,12 @@ $$latex literal$$
 """
 function break_me_docs end
 
+
+# `hasdoc` returns `true` on a name with a docstring.
+@test Docs.hasdoc(Base, :map)
+# `hasdoc` returns `false` on a name without a docstring.
+@test !Docs.hasdoc(Base, :_this_name_doesnt_exist_)
```
Comment (stevengj) [inline test/docs.jl:76]: Would be good to add a test for a name that does exist but isn't documented, e.g. 
```suggestion
@test !Docs.hasdoc(Base, :_this_name_doesnt_exist_)
@test isdefined(Base, :_typed_vcat) && !Docs.hasdoc(Base, :_typed_vcat)
```

**Comment by stevengj** [comment]
Needs a NEWS.md item.

File: NEWS.md [hunk 3]
```diff
@@ -61,6 +61,7 @@ New library features
 * `replace(string, pattern...)` now supports an optional `IO` argument to
   write the output to a stream rather than returning a string ([#48625]).
 * `sizehint!(s, n)` now supports an optional `shrink` argument to disable shrinking ([#51929]).
+* New function `Docs.hasdoc(module, symbol)::Bool` tells whether a name has a docstring ([#52139]).
```
Comment (stevengj) [inline NEWS.md:64]:
```suggestion
* New function `Docs.hasdoc(module, symbol)` tells whether a name has a docstring ([#52139]).
```

**Comment by jariji** [comment]
I don't know why it's failing.

**Comment by stevengj** [comment]
The CI failures look like unrelated glitches in the test system.

**Review by stevengj — APPROVED** [review]

**Comment by vtjnash** [comment]
Seems good to go

**Comment by LilithHafner** [comment]
Late to the party here, but wouldn't `has_docstring` be a more appropriate name as that is what this function actually checks? (a function may be documented via a mechanism other than a docstring)

Also this function should be public.

**Comment by jariji** [comment]
> a function may be documented via a mechanism other than a docstring

You mean it might be in the manual somewhere? I don't think I'd call a function "documented" if it doesn't have a docstring. Maybe others disagree?

**Comment by LilithHafner** [comment]
I was thinking about the other direction as well. If a function has a docstring but is not in the manual then, in Julia base at least, it is not considered "documented". E.g. the 1.10 API is defined in part with 

>  Functions, types, and constants are not part of the public API if they are not included in the documentation, _even if they have docstrings_.

**Comment by jariji** [comment]
I'd kinda forgotten about that policy. Tbh I don't see the advantage of it. 

I would think if a function is `public` it should be considered part of the API. I tend to think the manual is secondary to the experience of living in the runtime and using `?` --- if it points me to a manual page, I'll read it for context, but I don't think it's good to have to check an external manual to know whether a `public` function-with-docstring is "documented" or "public" so I can rely on it. I'd prefer a policy where I can check the API status of a function easily from within the system.

**Comment by DilumAluthge** [comment]
> Late to the party here, but wouldn't `has_docstring` be a more appropriate name as that is what this function actually checks? (a function may be documented via a mechanism other than a docstring)

I also think that this should be named `has_docstring`, to reduce ambiguity.

**Comment by DilumAluthge** [comment]
> Late to the party here, but wouldn't `has_docstring` be a more appropriate name as that is what this function actually checks? (a function may be documented via a mechanism other than a docstring)

https://github.com/JuliaLang/julia/pull/52724

**Comment by stevengj** [comment]
> Late to the party here, but wouldn't `has_docstring` be a more appropriate name as that is what this function actually checks?

I disagree, for reasons stated in https://github.com/JuliaLang/julia/pull/52724#issuecomment-1876137150 — this is in the `Docs` module, and in that context "doc" always refers to docstrings.

A `Documenter.hasdoc` function in the Documenter.jl module might check the manual, in contrast.

**Pull request merged**